engine/apps/labels/models.py

```diff
@@ -0,0 +1,112 @@
+import typing
+
+from django.db import models
+from django.utils import timezone
+
+from apps.labels.tasks import update_labels_cache
+from apps.labels.utils import LABEL_OUTDATED_TIMEOUT_MINUTES, LabelsData
+
+if typing.TYPE_CHECKING:
+    from apps.user_management.models import Organization
+
+
+class LabelKeyCache(models.Model):
+    id = models.CharField(primary_key=True, editable=False, max_length=36)
+    repr = models.CharField(max_length=200)
+    organization = models.ForeignKey("user_management.Organization", on_delete=models.CASCADE)
+    last_synced = models.DateTimeField(auto_now=True)
+
+    @property
+    def is_outdated(self) -> bool:
+        return timezone.now() - self.last_synced > timezone.timedelta(minutes=LABEL_OUTDATED_TIMEOUT_MINUTES)
+
+
+class LabelValueCache(models.Model):
+    id = models.CharField(primary_key=True, editable=False, max_length=36)
+    repr = models.CharField(max_length=200)
+    key = models.ForeignKey("labels.LabelKeyCache", on_delete=models.CASCADE, related_name="values")
+    last_synced = models.DateTimeField(auto_now=True)
+
+    @property
+    def is_outdated(self) -> bool:
+        return timezone.now() - self.last_synced > timezone.timedelta(minutes=LABEL_OUTDATED_TIMEOUT_MINUTES)
+
+
+class AssociatedLabel(models.Model):
+    """
+    Abstract model, is used to keep information about label association with other instances
+    (integrations, schedules, etc.). To add ability to associate labels with a type of instances ,
+    inhere this model and add a foreign key to the instance model.
+
+    Attention: add `AssociatedLabel` to the end of the name of inheritor (example: AlertReceiveChannelAssociatedLabel)
+    """
+
+    key = models.ForeignKey(LabelKeyCache, on_delete=models.CASCADE)
+    value = models.ForeignKey(LabelValueCache, on_delete=models.CASCADE)
+    organization = models.ForeignKey("user_management.Organization", on_delete=models.CASCADE, related_name="labels")
+
+    class Meta:
+        abstract = True
+
+    @staticmethod
+    def update_association(labels_data: "LabelsData", instance: models.Model, organization: "Organization") -> None:
```

Comment:
Any reason this is not an instance method instead of a static one? (so we can do `instance.update_association`)

Reply:
The `instance` here is the model instance that the labels are associated with (e.g. AlertReceiveChannel), not AssociatedLabel instance